src/Interpreters/Context.cpp

```diff
@@ -1393,62 +1395,128 @@ StoragePtr Context::executeTableFunction(const ASTPtr & table_expression, const
+                    }
+                    if (!structure_hint.empty())
+                    {
+                        if (use_structure_from_insertion_table_in_table_functions == 1)
+                            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
+
                         use_columns_from_insert_query = false;
                         break;
                     }
-                }
 
-                /// Check that all identifiers are column names from insertion table and not virtual column names from storage.
-                for (const auto & column_name : columns_names)
+                    asterisk = true;
+                }
+                else if (auto * function = (*expression)->as<ASTFunction>())
                 {
-                    if (!structure_hint.has(column_name) || virtual_column_names.contains(column_name))
+                    if (use_structure_from_insertion_table_in_table_functions == 2 && find_identifier(function))
                     {
                         use_columns_from_insert_query = false;
                         break;
                     }
-                }
 
-                /// If we don't have asterisk but only subset of columns, we should use
-                /// structure from insertion table only in case when table function
-                /// supports reading subset of columns from data.
-                if (use_columns_from_insert_query && !have_asterisk && !columns_names.empty())
+                    if (asterisk)
+                        --insert_structure_end;
+                    else
+                        ++insert_column;
+                }
+                else
                 {
-                    /// For input function we should check if input format supports reading subset of columns.
-                    if (table_function_ptr->getName() == "input")
-                        use_columns_from_insert_query = FormatFactory::instance().checkIfFormatSupportsSubsetOfColumns(getInsertFormat());
+                    if (asterisk)
+                        --insert_structure_end;
                     else
-                        use_columns_from_insert_query = table_function_ptr->supportsReadingSubsetOfColumns();
+                        ++insert_column;
                 }
             }
 
             if (use_columns_from_insert_query)
-                table_function_ptr->setStructureHint(structure_hint);
+            {
+                if (expression == expression_list.end())
+                {
+                    if (asterisk)
+                    {
+                        for (; insert_column != insert_structure_end; ++insert_column)
+                            structure_hint.add({ insert_column->name, insert_column->type });
+                    }
+
+                    if (!structure_hint.empty())
+                        table_function_ptr->setStructureHint(structure_hint);
```

Comment:
I still think when we have `auto` we should not set a hint when we have a list of column names and format that doesn't support reading subset of columns.

Reply:
but column list can be a full list of columns - this approach gives more flexibility - for example for csv with 2 columns we can have `SELECT c1, _path, c2`

Reply:
> this approach gives more flexibility 

Yes, that's true. But there is always a possibility that there is not a full list of columns, and in `auto` I would prefer the way that will always work. With schema inference it will always work. 

Anyway, that's a rare case I think, we can leave current way if you prefer it more.

Reply:
>in auto I would prefer the way that will always work

yeah, it makes sense... will fix this